File: internal/mmr/node.go
// Package mmr implements a Merkle Mountain Range (MMR), an append-only
// authenticated data structure for proving document state at any historical point.
package mmr

import (
	"crypto/sha256"
	"encoding/binary"
)

// DomainSeparators prevent second-preimage attacks by ensuring leaf hashes
// and internal node hashes are computed in distinct domains.
const (
	LeafPrefix     byte = 0x00
	InternalPrefix byte = 0x01
)

// HashSize is the size of SHA-256 output in bytes.
const HashSize = 32

// Node represents a single node in the Merkle Mountain Range.
type Node struct {
	Index  uint64   // Position in the MMR (0-indexed)
	Height uint8    // Height in the tree (0 = leaf)
	Hash   [32]byte // SHA-256 hash of the node
}

// NewLeafNode creates a leaf node from raw data.
// The hash is computed as SHA256(LeafPrefix || data) to prevent
// second-preimage attacks where internal nodes could be confused with leaves.
func NewLeafNode(index uint64, data []byte) *Node {
	h := sha256.New()
	h.Write([]byte{LeafPrefix})
	h.Write(data)

	var hash [32]byte
	copy(hash[:], h.Sum(nil))

	return &Node{
		Index:  index,
		Height: 0,
		Hash:   hash,
	}
}

// NewInternalNode creates an internal node by hashing two child nodes.
// The hash is computed as SHA256(InternalPrefix || leftHash || rightHash).
func NewInternalNode(index uint64, height uint8, left, right *Node) *Node {
	h := sha256.New()
	h.Write([]byte{InternalPrefix})
	h.Write(left.Hash[:])
	h.Write(right.Hash[:])

	var hash [32]byte
	copy(hash[:], h.Sum(nil))

	return &Node{
		Index:  index,
		Height: height,
		Hash:   hash,
	}
}

// HashLeaf computes the domain-separated hash of leaf data.
func HashLeaf(data []byte) [32]byte {
	h := sha256.New()
	h.Write([]byte{LeafPrefix})
	h.Write(data)

	var hash [32]byte
	copy(hash[:], h.Sum(nil))
	return hash
}

// HashInternal computes the domain-separated hash of two child hashes.
func HashInternal(left, right [32]byte) [32]byte {
	h := sha256.New()
	h.Write([]byte{InternalPrefix})
	h.Write(left[:])
	h.Write(right[:])

	var hash [32]byte
	copy(hash[:], h.Sum(nil))
	return hash
}

// Serialize converts a node to its binary representation for storage.
// Format: [8-byte Index][1-byte Height][32-byte Hash] = 41 bytes total
func (n *Node) Serialize() []byte {
	buf := make([]byte, 41)
	binary.BigEndian.PutUint64(buf[0:8], n.Index)
	buf[8] = n.Height
	copy(buf[9:41], n.Hash[:])
	return buf
}

// DeserializeNode reconstructs a node from its binary representation.
func DeserializeNode(data []byte) (*Node, error) {
	if len(data) < 41 {
		return nil, ErrInvalidNodeData
	}

	var hash [32]byte
	copy(hash[:], data[9:41])

	return &Node{
		Index:  binary.BigEndian.Uint64(data[0:8]),
		Height: data[8],
		Hash:   hash,
	}, nil
}

// NodeSize is the serialized size of a single node in bytes.
const NodeSize = 41

File: rust/witnessd-core/src/mmr/node.rs
use crate::mmr::errors::MmrError;
use sha2::{Digest, Sha256};

pub const HASH_SIZE: usize = 32;
pub const NODE_SIZE: usize = 41;

const LEAF_PREFIX: u8 = 0x00;
const INTERNAL_PREFIX: u8 = 0x01;

#[derive(Debug, Clone)]
pub struct Node {
    pub index: u64,
    pub height: u8,
    pub hash: [u8; HASH_SIZE],
}

pub fn hash_leaf(data: &[u8]) -> [u8; HASH_SIZE] {
    let mut hasher = Sha256::new();
    hasher.update([LEAF_PREFIX]);
    hasher.update(data);
    let digest = hasher.finalize();
    let mut out = [0u8; HASH_SIZE];
    out.copy_from_slice(&digest);
    out
}

pub fn hash_internal(left: [u8; HASH_SIZE], right: [u8; HASH_SIZE]) -> [u8; HASH_SIZE] {
    let mut hasher = Sha256::new();
    hasher.update([INTERNAL_PREFIX]);
    hasher.update(left);
    hasher.update(right);
    let digest = hasher.finalize();
    let mut out = [0u8; HASH_SIZE];
    out.copy_from_slice(&digest);
    out
}

impl Node {
    pub fn new_leaf(index: u64, data: &[u8]) -> Self {
        Self {
            index,
            height: 0,
            hash: hash_leaf(data),
        }
    }

    pub fn new_internal(index: u64, height: u8, left: &Node, right: &Node) -> Self {
        Self {
            index,
            height,
            hash: hash_internal(left.hash, right.hash),
        }
    }

    pub fn serialize(&self) -> [u8; NODE_SIZE] {
        let mut buf = [0u8; NODE_SIZE];
        buf[0..8].copy_from_slice(&self.index.to_be_bytes());
        buf[8] = self.height;
        buf[9..].copy_from_slice(&self.hash);
        buf
    }

    pub fn deserialize(data: &[u8]) -> Result<Self, MmrError> {
        if data.len() < NODE_SIZE {
            return Err(MmrError::InvalidNodeData);
        }
        let mut hash = [0u8; HASH_SIZE];
        hash.copy_from_slice(&data[9..41]);
        Ok(Self {
            index: u64::from_be_bytes(data[0..8].try_into().unwrap()),
            height: data[8],
            hash,
        })
    }
}
